DOC: Add replite console to the users docs

martinRenou opened the pull request:
## PR Summary

cc. @ianhi 

Add a JupyterLite console to the docs, allowing users to try and play with Matplotlib directly in the documentation 

![replite](https://user-images.githubusercontent.com/21197331/157636759-1fdbc20b-11e5-46b0-8a59-701dfd3f6d89.png)

I have not been able to build the docs locally due to some latex issue I'm having. But the CI should allow us to test this :)

martinRenou commented:
Turning this into draft until it actually works

ianhi commented:
:star_struck: :star_struck: :star_struck: 

r> I have not been able to build the docs locally due to some latex issue I'm having. 

They're very difficult to build locally - also takes a long time to build. In general I'd say it's not worth it. But if you do want to (and are on linux) then I found I was able to successfully do it by following this dockerfile: https://github.com/matplotlib/mpl-docker/blob/master/Dockerfile (or I usppose you could use that dockerfile)

martinRenou commented:
Thanks @ianhi! I'll see if I'm more lucky with this dockerfile, otherwise I'll rely on the CI

ianhi commented:
There is also this: https://matplotlib.org/stable/devel/dependencies.html#additional-external-dependencies

ianhi commented:
ooooh - @martinRenou you know what, a PR to [sphinx-gallery](https://github.com/sphinx-gallery/sphinx-gallery) enabling automatically interactive examples would probably be the most high impact jupyterlite docs PR you could make. Because then all https://matplotlib.org/stable/gallery/index would suddenly be runnable in the browser

ianhi commented:
See also: https://github.com/sphinx-gallery/sphinx-gallery/issues/927

tacaswell commented:
Building this locally and using `python -m http.server` to serve it I get 404s for 

```
http://localhost:8000/users/lite/repl/index.html?kernel=python&code=%20import%20matplotlib.pyplot%20as%20plt%20import%20numpy%20as%20np%20x%20=%20np.linspace(0,%202%20*%20np.pi,%20200)%20y%20=%20np.sin(x)%20fig,%20ax%20=%20plt.subplots()%20ax.plot(x,%20y)%20plt.show()
```

It looks like the issue is that the what ever is formatting the URL expects to be at the top level, but is actually a level down.

tacaswell commented:
Changing the URL in the iframe to `src="/lite/repl/index.html?kernel=python&code= import matplotlib.pyplot as plt import numpy as np x = np.linspace(0, 2 * np.pi, 200) y = np.sin(x) fig, ax = plt.subplots() ax.plot(x, y) plt.show()"` (by poking at it interactively in the browser) I can get it to load :tada:, however then it has the issue that the new lines are not preserved so it gives you a syntax error.

![image](https://user-images.githubusercontent.com/199813/157743141-3c2c856a-5b9c-4e39-9f9f-b17483561b94.png)

jtpio commented:
@tacaswell it should be possible to pass multiple `&code=` params, each corresponding to one line.

martinRenou commented:
@jtpio is there a way to keep the line break so that the whole code is in just one cell?

jtpio commented:
Normally it can be encoded and contain new lines. For example with the demo site:

https://jupyterlite.github.io/demo/repl/?kernel=python&toolbar=1&code=import%20matplotlib.pyplot%20as%20plt%0Aimport%20numpy%20as%20np%0Ax%20=%20np.linspace(0,%202%20*%20np.pi,%20200)%0Ay%20=%20np.sin(x)%0Afig,%20ax%20=%20plt.subplots()%0Aax.plot(x,%20y)%0Aplt.show()

![image](https://user-images.githubusercontent.com/591645/157908687-f3c9436e-4242-470b-a051-471ad8fa79fa.png)

martinRenou commented:
Thanks! I'll push a PR fixing jupyterlite-sphinx's replite directive

martinRenou commented:
Done in https://github.com/jupyterlite/jupyterlite-sphinx/pull/30

martinRenou commented:
I was finally able to build the Matplotlib docs locally :) With the new jupyterlite-sphinx version it works like a charm!

![matplotlib](https://user-images.githubusercontent.com/21197331/158164840-0f3ccf82-2a23-40bd-9373-f94e83d96d92.png)

Making this PR ready to review! Please tell me if you want any changes to be made

martinRenou commented:
I fail to understand why the CI is failing to build de docs. Though I can reproduce locally, `make html` fails (without a meaningful error message) while calling `sphinx-build . build` works correctly.

tacaswell commented:
There are many warnings that look like:

```
WARNING: multiple files found for the document "gallery/animation/animate_decay": ['gallery/animation/animate_decay.py', 'gallery/animation/animate_decay.py.md5', 'gallery/animation/animate_decay.rst', 'gallery/animation/animate_decay.ipynb']
Use '/home/circleci/project/doc/gallery/animation/animate_decay.rst' for the build.
WARNING: multiple files found for the document "gallery/animation/animated_histogram": ['gallery/animation/animated_histogram.py', 'gallery/animation/animated_histogram.py.md5', 'gallery/animation/animated_histogram.rst', 'gallery/animation/animated_histogram.ipynb']
Use '/home/circleci/project/doc/gallery/animation/animated_histogram.rst' for the build.
WARNING: multiple files found for the document "gallery/animation/animation_demo": ['gallery/animation/animation_demo.py', 'gallery/animation/animation_demo.py.md5', 'gallery/animation/animation_demo.rst', 'gallery/animation/animation_demo.ipynb']
Use '/home/circleci/project/doc/gallery/animation/animation_demo.rst' for the build.
WARNING: multiple files found for the document "gallery/animation/bayes_update": ['gallery/animation/bayes_update.py', 'gallery/animation/bayes_update.py.md5', 'gallery/animation/bayes_update.rst', 'gallery/animation/bayes_update.ipynb']
Use '/home/circleci/project/doc/gallery/animation/bayes_update.rst' for the build.
WARNING: multiple files found for the document "gallery/animation/unchained": ['gallery/animation/unchained.py.md5', 'gallery/animation/unchained.ipynb', 'gallery/animation/unchained.rst', 'gallery/animation/unchained.py']
Use '/home/circleci/project/doc/gallery/animation/unchained.rst' for the build.
WARNING: multiple files found for the document "gallery/animation/double_pendulum": ['gallery/animation/double_pendulum.py', 'gallery/animation/double_pendulum.py.md5', 'gallery/animation/double_pendulum.rst', 'gallery/animation/double_pendulum.ipynb']
Use '/home/circleci/project/doc/gallery/animation/double_pendulum.rst' for the build.
WARNING: multiple files found for the document "gallery/animation/dynamic_image": ['gallery/animation/dynamic_image.py', 'gallery/animation/dynamic_image.py.md5', 'gallery/animation/dynamic_image.rst', 'gallery/animation/dynamic_image.ipynb']
Use '/home/circleci/project/doc/gallery/animation/dynamic_image.rst' for the build.
WARNING: multiple files found for the document "gallery/animation/frame_grabbing_sgskip": ['gallery/animation/frame_grabbing_sgskip.py', 'gallery/animation/frame_grabbing_sgskip.rst', 'gallery/animation/frame_grabbing_sgskip.ipynb']
Use '/home/circleci/project/doc/gallery/animation/frame_grabbing_sgskip.rst' for the build.
WARNING: multiple files found for the document "gallery/animation/pause_resume": ['gallery/animation/pause_resume.py', 'gallery/animation/pause_resume.py.md5', 'gallery/animation/pause_resume.rst', 'gallery/animation/pause_resume.ipynb']
Use '/home/circleci/project/doc/gallery/animation/pause_resume.rst' for the build.
WARNING: multiple files found for the document "gallery/animation/rain": ['gallery/animation/rain.py', 'gallery/animation/rain.py.md5', 'gallery/animation/rain.rst', 'gallery/animation/rain.ipynb']
```

It looks like repllite is dropping some additional files in places that is confusing sphinx?

ianhi commented:
Is it possible to add ipympl to the jupyterlite environment here? This could easily be many peoples first interaction with matplotlib and it would be a great opportunity to reinforce the fact that matplotlib is in fact interactive. Thinking about this this because I saw: https://twitter.com/blu3r4d0n/status/1502309909811109888

martinRenou commented:
It would be great indeed @ianhi. Should we probably wait for ipywidgets to work in JupyterLab consoles though? So that people don't get confused as to why ipympl doesn't work in the replite. @jtpio had some ideas on improving the replite, making it so that widgets could probably work as well.

martinRenou commented:
Not difficult :) We just need to change the directive from `replite` to `retrolite`, and we probably need to add a Notebook to the docs.

To summarize, what's remaining before the PR is working and ready:
- [ ] : Understand/Fix the CI issue
- [ ] : Add ipympl to the deployment, and use retrolite for the Notebook
- [ ] : Fix https://github.com/jupyterlite/jupyterlite-sphinx/issues/36

martinRenou commented:
> It looks like repllite is dropping some additional files in places that is confusing sphinx?

What really confuses me is that jupyterlite did not even run yet at the point the warnings are showing up. And those files seems to be more related to the Matplotlib docs than JupyterLite? But the weird thing is that those warnings don't seem to show up in other PRs.

timhoffm commented:
This is indeed somewhat strange. Maybe remove the `.. replite::` for a test? It seems unlikely, but perhaps it somehow messes up the ReST structure or sphinx internals.

SylvainCorlay commented:
I would love to see this go through!

martinRenou commented:
This is a fair point 👍🏽 I will change the PR so that it does not include the repl in the main page, instead the user will need to click a button to navigate to `/lite/lab` in order to get the full JupyterLab deployment. That way they can use Matplotlib and the ipympl widget there, but the user docs is not polluted.

jklymak commented:
Thanks @martinRenou that would be my preference, based on my loose understanding of how this works.  But its possible others are less conservative and are for this...

martinRenou commented:
@jklymak Slight change in the plan, I pushed some changes to jupyterlite-sphinx so that it lazy-loads the JupyterLite UI (only load upon user click). This way it does not make the docs heavier for those who don't want to start the widget, but we still can embed it:

(don't mind the kernel selection I just need to tweak replite properly to get rid of it)
![repl](https://user-images.githubusercontent.com/21197331/181029735-05141415-f48e-48d7-9511-fc519c389bcf.gif)

Also, the user will have access of the entire gallery of Notebooks in the JupyterLab deployment that comes with it:

![gallery](https://user-images.githubusercontent.com/21197331/181030407-18204351-ff69-4420-b3bf-ff2ce0a6c413.png)

I will push some more tweaks, but comments are very welcome :D

jklymak commented:
But maybe this belongs on the brochure page versus the doc page?  Also the font cache rebuilding each startup is suboptimal.  Does that get cached at least?

martinRenou commented:
> Also the font cache rebuilding each startup is suboptimal. Does that get cached at least?

I suppose no. Maybe we can change the way Matplotlib is packaged on emscripten-forge https://github.com/emscripten-forge/recipes/tree/main/recipes/recipes_emscripten/matplotlib so that the fonts are part of the output?

> But maybe this belongs on the brochure page versus the doc page?

Sorry, what do you mean by brochure page?

jklymak commented:
https://matplotlib.org is at https://github.com/matplotlib/mpl-brochure-site. This is currently targeting the top-level of the docs, versus the top level of the org.  I don't feel strongly that it cannot go at the top level of the docs, but it seems maybe more at home at the top level?

martinRenou commented:
@jklymak One problem I see with moving the PR to https://github.com/matplotlib/mpl-brochure-site is that the user will not have access to the gallery Notebooks in the JupyterLite deployment. Or am I missing something?

tacaswell commented:
I am happy to put this in the per-version docs to start with.  I think we will eventually want to have a version on https://matplotlib.org but I think it will be an "and"  not an "or".

It is pretty slick that we can just expose the fully gallery like this!

Is it possible to put this behind a feature flag when building the docs?  This is both to reduce the burden on down-stream packagers who might need to packaging the repllite dependencies for their platforms if they want to build the docs and it is hard enough to build the docs locally.

tacaswell commented:
Trying to build this locally I'm having issues with the lite/lab part:

```
FileNotFoundError: [Errno 44] No such file or directory: '/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/matplotlib/mpl-data/matplotlibrc'

```

martinRenou commented:
Thank you @tacaswell for testing! At what stage are you seeing this error?

tacaswell commented:
Everything builds fine, but when I try to actually use the repllite I get

```
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
/tmp/xpython_42/3500927980.py in <cell line: 1>()
----> 1 get_ipython().run_line_magic('matplotlib', 'inline')
      2 
      3 import matplotlib.pyplot as plt
      4 import numpy as np
      5 

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/IPython/core/interactiveshell.py in run_line_magic(self, magic_name, line, _stack_depth)
   2405                 kwargs['local_ns'] = self.get_local_scope(stack_depth)
   2406             with self.builtin_trap:
-> 2407                 result = fn(*args, **kwargs)
   2408             return result
   2409 

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/decorator.py in fun(*args, **kw)
    230             if not kwsyntax:
    231                 args, kw = fix(args, kw, sig)
--> 232             return caller(func, *(extras + args), **kw)
    233     fun.__name__ = func.__name__
    234     fun.__doc__ = func.__doc__

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/IPython/core/magic.py in <lambda>(f, *a, **k)
    185     # but it's overkill for just that one bit of state.
    186     def magic_deco(arg):
--> 187         call = lambda f, *a, **k: f(*a, **k)
    188 
    189         if callable(arg):

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/IPython/core/magics/pylab.py in matplotlib(self, line)
     97             print("Available matplotlib backends: %s" % backends_list)
     98         else:
---> 99             gui, backend = self.shell.enable_matplotlib(args.gui.lower() if isinstance(args.gui, str) else args.gui)
    100             self._show_matplotlib_backend(args.gui, backend)
    101 

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/IPython/core/interactiveshell.py in enable_matplotlib(self, gui)
   3598         """
   3599         from IPython.core import pylabtools as pt
-> 3600         from matplotlib_inline.backend_inline import configure_inline_support
   3601         gui, backend = pt.find_gui_and_backend(gui, self.pylab_gui_select)
   3602 

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/matplotlib_inline/backend_inline.py in <module>
      4 # Distributed under the terms of the BSD 3-Clause License.
      5 
----> 6 import matplotlib
      7 from matplotlib.backends.backend_agg import (  # noqa
      8     new_figure_manager,

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/matplotlib/__init__.py in <module>
    874 # by explicitly calling the superclass (dict.update, dict.items) to avoid
    875 # triggering resolution of _auto_backend_sentinel.
--> 876 rcParamsDefault = _rc_params_in_file(
    877     cbook._get_data_path("matplotlibrc"),
    878     # Strip leading comment.

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/matplotlib/__init__.py in _rc_params_in_file(fname, transform, fail_on_error)
    776     import matplotlib as mpl
    777     rc_temp = {}
--> 778     with _open_file_or_url(fname) as fd:
    779         try:
    780             for line_no, line in enumerate(fd, 1):

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/contextlib.py in __enter__(self)
    133         del self.args, self.kwds, self.func
    134         try:
--> 135             return next(self.gen)
    136         except StopIteration:
    137             raise RuntimeError("generator didn't yield") from None

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/matplotlib/__init__.py in _open_file_or_url(fname)
    753         if encoding is None:
    754             encoding = "utf-8"
--> 755         with open(fname, encoding=encoding) as f:
    756             yield f
    757 

FileNotFoundError: [Errno 44] No such file or directory: '/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/matplotlib/mpl-data/matplotlibrc'
```

on trying to import Matplotlib.

At some point recently we adjusted how we did the rcparams to directly read the template file so that it would not get out-of-sync between the various places.   My suspicion is that either the process that is bundling everything up is missing the file (which may be an issue in our manifest?) or we need to find some otherway to read that file in at import time (like stuff it in a module...it is just a very long string).

martinRenou commented:
This is unfortunate. 

An new `empack` release was out, and it filters out the `matplotlibrc` file which means it's not included in the xeus-python runtime. So this is not on your end. I will try a fix today.

tacaswell commented:
I tried to install from the empack main branch but now my build fails with

```
Transaction finished
[LiteBuildApp] ERROR | [lite] [post_build] [jupyterlite-xeus-python] [ERR] 2 validation errors for FileFilter
include_patterns -> 0 -> __root__ -> pattern
  extra fields not permitted (type=value_error.extra)
include_patterns -> 0 -> __root__ -> regex
  extra fields not permitted (type=value_error.extra)
Traceback (most recent call last):
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/doit/doit_cmd.py", line 294, in run
    return command.parse_execute(args)
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/doit/cmd_base.py", line 150, in parse_execute
    return self.execute(params, args)
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/doit/cmd_base.py", line 570, in execute
    return self._execute(**exec_params)
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/doit/cmd_run.py", line 265, in _execute
    return runner.run_all(self.control.task_dispatcher())
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/doit/runner.py", line 254, in run_all
    self.run_tasks(task_dispatcher)
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/doit/runner.py", line 213, in run_tasks
    node = task_dispatcher.generator.send(node)
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/doit/control.py", line 629, in _dispatcher_generator
    next_step = node.step()
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/doit/control.py", line 336, in step
    return next(self.generator)
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/doit/control.py", line 345, in _func
    for value in decorated(*args, **kwargs):
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/doit/control.py", line 473, in _add_task
    new_tasks = generate_tasks(to_load, task_gen, ref.__doc__)
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/doit/loader.py", line 390, in generate_tasks
    for task_dict, x_doc in flat_generator(gen_result, gen_doc):
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/doit/loader.py", line 27, in flat_generator
    for item in gen:
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/jupyterlite/manager.py", line 134, in _delayed_gather
    for task in _gather():
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/jupyterlite/manager.py", line 127, in _gather
    raise err
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/jupyterlite/manager.py", line 117, in _gather
    for task in getattr(addon, attr)(self):
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/jupyterlite_xeus_python/env_build_addon.py", line 127, in post_build
    pack_python_core(
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/empack/file_packager.py", line 213, in pack_python_core
    pack_environment(
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/empack/file_packager.py", line 186, in pack_environment
    filter_env(env_prefix=env_prefix, target_dir=target_dir)
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/empack/filter_env.py", line 74, in filter_env
    filter_pkg(
  File "/home/tcaswell/.virtualenvs/dd310/lib/python3.10/site-packages/empack/filter_env.py", line 49, in filter_pkg
    include_patterns = FileFilter.parse_obj(pack_config["packages"][name])
  File "pydantic/main.py", line 521, in pydantic.main.BaseModel.parse_obj
  File "pydantic/main.py", line 341, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 2 validation errors for FileFilter
include_patterns -> 0 -> __root__ -> pattern
  extra fields not permitted (type=value_error.extra)
include_patterns -> 0 -> __root__ -> regex
  extra fields not permitted (type=value_error.extra)
[LiteBuildApp] Exiting application: jupyter

Extension error (jupyterlite_sphinx.jupyterlite_sphinx):
Handler <function jupyterlite_build at 0x7fe485b79bd0> for event 'build-finished' threw an exception (exception: Command '['jupyter', 'lite', 'build', '--debug', '--config', 'jupyter_lite_config.json', '--contents', 'gallery', '--contents', '/home/tcaswell/source/p/matplotlib/matplotlib/doc/_contents', '--output-dir', '/home/tcaswell/source/p/matplotlib/matplotlib/doc/build/html/lite', '--lite-dir', '.']' returned non-zero exit status 3.)
make: *** [Makefile:34: html] Error 2
make: Leaving directory '/home/tcaswell/source/p/matplotlib/matplotlib/doc'

```

martinRenou commented:
Thanks for testing @tacaswell ! We'll make sure to make it work and make a patch release.

There have been multiple changes in the way we filter site-packages (we have to filter the site-packages content so that xeus-python-lite don't get crazy big to download for users)

martinRenou commented:
@tacaswell It should now work again with the empack 0.8.2 release.

I will look into pre-packaging the font cache so that Matplotlib does not have to build the cache at startup.

tacaswell commented:
That works now, however if I try to use ipympl

```
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
/tmp/xpython_42/4022676443.py in <cell line: 1>()
----> 1 get_ipython().run_line_magic('matplotlib', 'widget')

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/IPython/core/interactiveshell.py in run_line_magic(self, magic_name, line, _stack_depth)
   2405                 kwargs['local_ns'] = self.get_local_scope(stack_depth)
   2406             with self.builtin_trap:
-> 2407                 result = fn(*args, **kwargs)
   2408             return result
   2409 

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/decorator.py in fun(*args, **kw)
    230             if not kwsyntax:
    231                 args, kw = fix(args, kw, sig)
--> 232             return caller(func, *(extras + args), **kw)
    233     fun.__name__ = func.__name__
    234     fun.__doc__ = func.__doc__

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/IPython/core/magic.py in <lambda>(f, *a, **k)
    185     # but it's overkill for just that one bit of state.
    186     def magic_deco(arg):
--> 187         call = lambda f, *a, **k: f(*a, **k)
    188 
    189         if callable(arg):

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/IPython/core/magics/pylab.py in matplotlib(self, line)
     97             print("Available matplotlib backends: %s" % backends_list)
     98         else:
---> 99             gui, backend = self.shell.enable_matplotlib(args.gui.lower() if isinstance(args.gui, str) else args.gui)
    100             self._show_matplotlib_backend(args.gui, backend)
    101 

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/IPython/core/interactiveshell.py in enable_matplotlib(self, gui)
   3611                 gui, backend = pt.find_gui_and_backend(self.pylab_gui_select)
   3612 
-> 3613         pt.activate_matplotlib(backend)
   3614         configure_inline_support(self, backend)
   3615 

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/IPython/core/pylabtools.py in activate_matplotlib(backend)
    357     from matplotlib import pyplot as plt
    358 
--> 359     plt.switch_backend(backend)
    360 
    361     plt.show._needmain = False

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/matplotlib/pyplot.py in switch_backend(newbackend)
    265     backend_name = cbook._backend_module_name(newbackend)
    266 
--> 267     class backend_mod(matplotlib.backend_bases._Backend):
    268         locals().update(vars(importlib.import_module(backend_name)))
    269 

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/matplotlib/pyplot.py in backend_mod()
    266 
    267     class backend_mod(matplotlib.backend_bases._Backend):
--> 268         locals().update(vars(importlib.import_module(backend_name)))
    269 
    270     required_framework = _get_required_interactive_framework(backend_mod)

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/importlib/__init__.py in import_module(name, package)
    124                 break
    125             level += 1
--> 126     return _bootstrap._gcd_import(name[level:], package, level)
    127 
    128 

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/importlib/_bootstrap.py in _gcd_import(name, package, level)

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/importlib/_bootstrap.py in _find_and_load(name, import_)

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/importlib/_bootstrap.py in _find_and_load_unlocked(name, import_)

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/importlib/_bootstrap.py in _load_unlocked(spec)

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/importlib/_bootstrap_external.py in exec_module(self, module)

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/importlib/_bootstrap.py in _call_with_frames_removed(f, *args, **kwds)

/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/ipympl/backend_nbagg.py in <module>
    117 
    118 here = os.path.dirname(__file__)
--> 119 with open(os.path.join(here, 'static', 'package.json')) as fid:
    120     js_version = json.load(fid)['version']
    121 

FileNotFoundError: [Errno 44] No such file or directory: '/tmp/xeus-python-kernel/envs/xeus-python-kernel/lib/python3.10/site-packages/ipympl/static/package.json'
```

tacaswell commented on doc/index.rst:
```diff
@@ -90,6 +89,27 @@ Learning resources
+   :height: 600px
+   :prompt: Try Matplotlib!
+
+   %matplotlib inline
```
```suggestion
   %matplotlib widget
```

tacaswell replied:
If we are going to embed an interactive section we should use a live backend!

martinRenou replied:
I totally agree, although widgets don't work in the replite UI unfortunately.

Alternatively we could show the retrolab UI instead, which does support widgets. But it's a more "crowded" UI. What do you think?

tacaswell replied:
I would go for an interactive backend over almost anything.

martinRenou replied:
Sounds good! I will push the require changes

martinRenou commented:
Thank you again for trying ipympl. A fix has been included, I need to make the proper changes to jupyterlite-xeus-python in order to include them.

jklymak commented:
but again if we go with a heavier backend, that is even more time for the server to push more code onto the users machine.   How much bandwidth does this all take, and at what stage?  I hope the first button is light, but that hitting it triggers downloading the rest?

martinRenou commented:
> I hope the first button is light, but that hitting it triggers downloading the rest?

Yes :) And the rest is cached by the browser, so if you download it once, it should be fine the next time you come.

But yes, if you don't click the first button, you don't download anything apart from the button HTML.

tacaswell commented:
We will also need to make sure that cloudflare is set up to cache all of these files so we do not blow up our hosting bandwidth with digital ocean!

martinRenou commented:
I updated jupyterlite-xeus-python with the latest versions that fixes ipympl support and remove the `Matplotlib is building font cache` warning (by pre-building the cache).

I also switched to the retrolab UI with an ipympl example instead :)

SylvainCorlay commented:
This is looking pretty good !

tacaswell commented:
Something is not right as hosted from circle (I could beleive this is a hosting / deployment issue) and the main example gives a 404 for https://output.circle-artifacts.com/output/job/a92f3656-f63b-4e25-98b0-fab5d09b2607/artifacts/0/doc/build/html/lite/retro/notebooks/?path=matplotlib.ipynb and https://output.circle-artifacts.com/output/job/a92f3656-f63b-4e25-98b0-fab5d09b2607/artifacts/0/doc/build/html/lite/lab

I still think that this needs a build flag so that down-stream packagers and developers can skip this as part of the building the docs.

martinRenou commented:
It looks like a redirection issue, if you look at https://output.circle-artifacts.com/output/job/a92f3656-f63b-4e25-98b0-fab5d09b2607/artifacts/0/doc/build/html/lite/lab you don't get properly redirected to https://output.circle-artifacts.com/output/job/a92f3656-f63b-4e25-98b0-fab5d09b2607/artifacts/0/doc/build/html/lite/lab/index.html

So maybe an issue on the circleCI deployment indeed? Also it is super slow to load the assets, are there bandwidth limitations on this deployment?

> I still think that this needs a build flag so that down-stream packagers and developers can skip this as part of the building the docs.

Yes! Any idea how to implement this? Where should this flag be?

QuLogic commented:
> So maybe an issue on the circleCI deployment indeed? Also it is super slow to load the assets, are there bandwidth limitations on this deployment?

It's just CircleCI's S3 bucket, so I would not expect it to be particularly slow, though it is slower generally. It seems to be loading every file serially; is that a replite bug? 

> > I still think that this needs a build flag so that down-stream packagers and developers can skip this as part of the building the docs.
> 
> Yes! Any idea how to implement this? Where should this flag be?

It can be configured by something like the release tag, or the release tag itself:
https://github.com/matplotlib/matplotlib/blob/main/doc/conf.py#L27

QuLogic commented:
Also, trying out your link, while it takes a couple minutes to load, I also get several script loading failures in the console, a `ServiceWorker` failure, and finally no Kernels available.

SylvainCorlay commented:
Is there any outstanding obstacle to getting this one in?

timhoffm commented:
Sorry for not being responsive. People are quite busy here and I think nobody is an expert to review in this topic.

I've restarted the docs build because the existing one had expired. The code looks fundamentally reasonable. To move this foraward, I suggest to merge this by judiging the visible result in the docs.

timhoffm commented:
When trying this on CI:
https://output.circle-artifacts.com/output/job/3481f984-d9fa-42f6-aab4-f029b2bd34ec/artifacts/0/doc/build/html/index.html

and clicking on the live box, I get:
![image](https://user-images.githubusercontent.com/2836374/194644618-44809910-91cf-49db-ac5b-ebd1138fec6e.png)

tacaswell commented:
Outstanding issues:

Can the extension be made read-parallel safe (is it already and we just need declare it?)?

> WARNING: the jupyterlite_sphinx extension does not declare if it is safe for parallel reading, assuming it isn't - please ask the extension author to check and make it explicit
WARNING: doing serial read

There is a veritable wall of errors that look like:

```
--- Logging error ---
Traceback (most recent call last):
  File "/home/tcaswell/.pybuild/dd311/lib/python3.11/logging/__init__.py", line 1110, in emit
    msg = self.format(record)
          ^^^^^^^^^^^^^^^^^^^
  File "/home/tcaswell/.pybuild/dd311/lib/python3.11/logging/__init__.py", line 953, in format
    return fmt.format(record)
           ^^^^^^^^^^^^^^^^^^
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/traitlets/config/application.py", line 141, in format
    return super().format(record)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tcaswell/.pybuild/dd311/lib/python3.11/logging/__init__.py", line 687, in format
    record.message = record.getMessage()
                     ^^^^^^^^^^^^^^^^^^^
  File "/home/tcaswell/.pybuild/dd311/lib/python3.11/logging/__init__.py", line 377, in getMessage
    msg = msg % self.args
          ~~~~^~~~~~~~~~~
TypeError: not all arguments converted during string formatting
Call stack:
  File "/home/tcaswell/.virtualenvs/dd311/bin/jupyter-lite", line 8, in <module>
    sys.exit(main())
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/jupyter_core/application.py", line 276, in launch_instance
    return super().launch_instance(argv=argv, **kwargs)
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/traitlets/config/application.py", line 982, in launch_instance
    app.start()
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/jupyter_core/application.py", line 265, in start
    self.subapp.start()
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/jupyterlite/app.py", line 217, in start
    super().start()
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/jupyterlite/app.py", line 169, in start
    self.exit(self.lite_manager.doit_run(*self._doit_cmd))
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/jupyterlite/manager.py", line 56, in doit_run
    return runner.run([task, *args])
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/doit/doit_cmd.py", line 294, in run
    return command.parse_execute(args)
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/doit/cmd_base.py", line 150, in parse_execute
    return self.execute(params, args)
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/doit/cmd_base.py", line 570, in execute
    return self._execute(**exec_params)
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/doit/cmd_run.py", line 265, in _execute
    return runner.run_all(self.control.task_dispatcher())
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/doit/runner.py", line 254, in run_all
    self.run_tasks(task_dispatcher)
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/doit/runner.py", line 220, in run_tasks
    base_fail = self.execute_task(node.task)
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/doit/runner.py", line 176, in execute_task
    return task.execute(self.stream)
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/doit/task.py", line 473, in execute
    action_return = action.execute(task_stdout, task_stderr)
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/doit/action.py", line 461, in execute
    returned_value = self.py_callable(*self.args, **kwargs)
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/jupyterlite/addons/contents.py", line 25, in <lambda>
    pprint.pformat([str(p[0]) for p in self.file_src_dest]),
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/jupyterlite/addons/contents.py", line 25, in <listcomp>
    pprint.pformat([str(p[0]) for p in self.file_src_dest]),
  File "/home/tcaswell/.virtualenvs/dd311/lib/python3.11/site-packages/jupyterlite/addons/contents.py", line 118, in file_src_dest
    self.log.debug("Already populated", resolved)
Message: 'Already populated'
Arguments: ('/home/tcaswell/source/p/matplotlib/matplotlib/doc/build/html/lite/files/annotation_polar.ipynb',)

```

It seems to result in a working website, but (locally) fails the build for me.

This seems to add 400K to the main page, but if you click the "try Matplotlib" button it is on the order of 64M and over my local network takes 8-9 s to load!  I think we need a warning and note about the size to expect when clicking the button. 

I think we need to make the "view the gallery" link needs to be _way_ more obvious.  I knew it was there and took me few seconds to find it.

Launching jupyterlab produces a warning about a git extension and costs about 12M and 10s or so.  This probably should also be warned?

The notebooks launch with `%matplotlib inline` at the top, is it possible to change this to `%matplotlib widget` ?

In jupyterlab there are directories full of .ipynb, but at the toplevel there is a mix of .ipynb, rst, and pickles (?!).

It seems every time a new notebook is opened another ~50-60M of data is pulled.  I'm not sure if this is an inherent problem or the cache being disabled by the dev tools (even if "disable cache" is unchecked?). 


I'm going to push a rebase.

agriyakhetarpal commented:
Hi all and to those subscribed to this PR, I've been interested in getting this effort revived for Matplotlib through https://github.com/jupyterlite/jupyterlite-sphinx/releases/tag/v0.18.0 that we've now released today on 13/01/2025. I've read through this thread fully, and I'm happy to share that most of the problems noticed in `jupyterlite-sphinx` and the JupyterLite/WASM ecosystem have been more or less resolved at this point :)

Here is a table that contains a short summary of relevant discussions and previous requests in the code review above for easy access, and my comments on them:

| Review comment | Link to review comment | My notes |
|--------|--------|--------|
| Request for Sphinx-Gallery to be able to run all examples interactively in the browser | https://github.com/matplotlib/matplotlib/pull/22634#issuecomment-1064315131 | ✅ This is now possible in many variations. Sphinx-Gallery offers a [JupyterLite integration](https://sphinx-gallery.github.io/stable/configuration.html#jupyterlite) via `jupyterlite-sphinx`, and [`jupyterlite-sphinx` itself offers a `TryExamples` directive](https://jupyterlite-sphinx.readthedocs.io/en/stable/directives/try_examples.html) to add interactivity to doctest-based examples residing in docstrings  |
| 404 error when opening the REPL | https://github.com/matplotlib/matplotlib/pull/22634#issuecomment-1065280938 | ✅ Fixed here itself and available in `jupyterlite-sphinx` 0.4.2 and later, shouldn't happen now  |
| `ipympl` availability | https://github.com/matplotlib/matplotlib/pull/22634#issuecomment-1066908684 | ⚠️ The WebAgg backend used `tornado` which was replaced with `asyncio` in https://github.com/matplotlib/matplotlib/pull/20591, things should work now but I'm not sure of its status |
| `ipywidgets` | https://github.com/matplotlib/matplotlib/pull/22634#issuecomment-1067709156 | ✅ Also available; a few regressions reported from time to time, but now stable and all resolved in early 2025 |
| Replacing the REPL | https://github.com/matplotlib/matplotlib/pull/22634#issuecomment-1068141366 | ✅ If the REPL is not needed, `jupyterlite-sphinx` also offers options for other directives, such as [JupyterLite](https://jupyterlite-sphinx.readthedocs.io/en/stable/directives/jupyterlite.html) and [NotebookLite](https://jupyterlite-sphinx.readthedocs.io/en/stable/directives/notebooklite.html) (that expose the Jupyter Lab and Notebook interfaces respectively), besides [`TryExamples`](https://jupyterlite-sphinx.readthedocs.io/en/stable/directives/try_examples.html) as mentioned above |
| Concerns around bandwidth consumptions as the REPL runs when the page loads | https://github.com/matplotlib/matplotlib/pull/22634#pullrequestreview-1021366696 | ✅ The [auto-execution of the REPL can now be disabled](https://jupyterlite-sphinx.readthedocs.io/en/stable/configuration.html#repl-code-auto-execution-with-the-replite-directive) globally and on a per-directive basis with `jupyterlite-sphinx` 0.18.0, based on the upstream change https://github.com/jupyterlite/jupyterlite/pull/1547, courtesy of @jtpio |
| CircleCI PR previews of the docs report broken links | https://github.com/matplotlib/matplotlib/pull/22634#issuecomment-1207471380 | ✅ This has been reported on other projects' documentation websites as well, such as scikit-learn, NumPy, and SciPy; all fixed in our previous 0.17 release |
| Request for a build-time flag for downstream packagers to disable JupyterLite when building the docs | https://github.com/matplotlib/matplotlib/pull/22634#issuecomment-1207471380 | 🤔 It should be possible to add a `disable_jupyterlite` boolean which can be set to `False` to pop `jupyterlite_sphinx` out of the list of extensions. Dependencies can be managed separately as well, if it's required to leave them out of `[docs]` extras and the like, of course. |
| Make `jupyterlite-sphinx` parallel-read safe | https://github.com/matplotlib/matplotlib/pull/22634#issuecomment-1363554668 | ✅ Implemented now since several past releases |
| Warnings about high bandwidth usage and high load times | https://github.com/matplotlib/matplotlib/pull/22634#issuecomment-1363554668 | ✅ Available since early 2024, the text can be configured. As a bonus, it's also possible to disable JupyterLite on mobile devices by removing the visual elements altogether with CSS, and I've been working on making that better: https://github.com/jupyterlite/jupyterlite-sphinx/pull/250 |

@martinRenou, I wish to ask if you're still interested in completing this PR? I would love to supersede it, but wished to ask in advance. If yes, no worries, I would be happy to step back and let you complete it. :)

Additional comments:

- As it was suggested in https://github.com/matplotlib/matplotlib/pull/22634#issuecomment-1201984679 that such a REPL would also make sense to add on Matplotlib's brochure site/landing page at https://matplotlib.org//, besides the docs. I think a button similar to the "Try Matplotlib (on binder)" button, but one that says "Try Matplotlib (JupyterLite" or similar could be added there, which would simply open a REPL or a Jupyter Lab environment with Matplotlib installed. I would be happy to take it up.
- Both the [Xeus](https://github.com/jupyterlite/xeus) and [Pyodide](https://github.com/jupyterlite/pyodide-kernel) kernels are reasonably mature and stable enough at this point to run Matplotlib without issues, but we should be fine with choosing either of them for this task (probably Xeus is better)

agriyakhetarpal commented:
Hi @martinRenou, I hope it is fine to ping you again here? Perhaps you missed my comment above or didn't see this PR show up in your notifications – I've been planning to work on this lately and hope to open a new PR to supersede this, and I have a working draft on my fork via https://github.com/agriyakhetarpal/matplotlib/pull/1 that preserves authorship. If you prefer I don't do this, please let me know, and thanks for your work!

martinRenou commented:
Hi @agriyakhetarpal! First, thanks a ton for your work in jupyterlite-sphinx 🤟🏽, thanks for the detailed comment here, and thanks for taking this PR over!

Happy to close this one and have yours replace it 😃

martinRenou commented:
Closing as replaced by https://github.com/matplotlib/matplotlib/pull/29506
